### PR TITLE
Additional http headers for websocket connection

### PR DESCRIPTION
**Pull request by KSDaemon**
- Added possibility to provide additional HTTP headers to the underlying websocket connection
- Additional ERROR message `AUTHENTICATION_FAILED`

### CODE REVIEW

**Comment by konsultaner** [comment]
I had a quick check. looks good. I'll give it a deep one in about 1.5h. Maybe I'll get to merge it today :)

**Comment by KSDaemon** [comment]
Great! Thnx! Looking forward it!)

File: lib/src/client.dart [hunk 1]
```diff
@@ -112,6 +113,7 @@ class Client {
           _connect(options);
         } else {
           _controller.addError(abort);
+          transport.onConnectionLost!.completeError('Closing');
```
Comment (konsultaner) [inline lib/src/client.dart:116]: Could you please provide a unit test case for this?

File: lib/src/transport/websocket/websocket_transport_io.dart [hunk 2]
```diff
@@ -75,7 +79,7 @@ class WebSocketTransport extends AbstractTransport {
     _onDisconnect = Completer();
     _onConnectionLost = Completer();
     try {
-      _socket = await WebSocket.connect(_url, protocols: [_serializerType]);
+      _socket = await WebSocket.connect(_url, protocols: [_serializerType], headers: _headers);
```
Comment (konsultaner) [inline lib/src/transport/websocket/websocket_transport_io.dart:82]: @KSDaemon Could you also add a test for the headers? maybe in the websocket io test?

Comment (KSDaemon) [reply lib/src/transport/websocket/websocket_transport_io.dart:82]: @konsultaner Thnx for the review! I'll cover this with tests of course!

Comment (KSDaemon) [reply lib/src/transport/websocket/websocket_transport_io.dart:82]: No problem! And btw I think this is the responsibility of the author of changes to cover them with tests!  ;)

**Comment by KSDaemon** [comment]
@konsultaner Hi! I've added tests. Pls have a look. It's a little bit ugly, but don't want to create more complex flow. But I adopt it if you suggest something else.

**Comment by KSDaemon** [comment]
Gentle reminder about PR :)

File: lib/src/client.dart (hunk 1 above)
Comment (KSDaemon) [reply lib/src/client.dart:116]: @konsultaner Can you help me with this check? Can not figure out how to check for connection loss after all...

**Comment by KSDaemon** [comment]
It took time but seems all is cleared now. Tests are passing. @konsultaner Can you have a look?

**Comment by konsultaner** [comment]
I was on a shorts holiday. Will have a look at it tomorrow morning 🤟

File: lib/src/client.dart [hunk 3]
```diff
@@ -84,24 +84,30 @@ class Client {
...
       try {
         var session = await Session.start(realm, transport,
             authId: authId,
             authRole: authRole,
             authMethods: authenticationMethods,
             reconnect: options.reconnectTime);
         _controller.add(session);
+
+        unawaited(transport.onConnectionLost!.future.then((_) async {
+            // Trying to reconnect only if we were connected and there is
```
Comment (konsultaner) [inline lib/src/client.dart:96]: I get the point, but what If the router crashes and auto heals. The there will be only one reconnect. I guess the point was to not reconnect if the credentials were wrong?

Comment (KSDaemon) [reply lib/src/client.dart:96]: Hi @konsultaner! Great that you get to this! 
Yeah, original idea was not to reconnect when it is not useful, like with bad credentials. 

The code in `client.dart` is a little bit complicated... Maybe we can refactor this in future :) 

So if you know how to resolve this problem in more elegant way — please suggest!

**Comment by konsultaner** [comment]
I really like that you considered a different reconnect strategy. I'm not sure this will work in an auto-healing environment. We may have to consider a more complex reconnect strategy depending on the reason why it didn't work or maybe have it configurable? What do you think @KSDaemon

**Comment by KSDaemon** [comment]
Yeah! Right now lib operates only on transport state to take decision: reconnect or not. But there are cases when session state must be taken into account. 

Can you describe a bit more about `auto-healing` envs? I mean that in general (and as I see it in some wamp clients/libs) when the transport connection is aborted/broken/closed — the session is finished. And new reconnection from client means getting new sessionId. Even if it auto resubscribes to topics, the respective subscriptionIds will be different.

**Comment by KSDaemon** [comment]
So in the first run, there are the next points of connection failure and reconnection:
* Transport failures before session is established.
* Wamp abort (due to auth or other reasons) before session is established. Wamp error means that server processed client messages and get to some point that resolves in about. Does auto reconnection fix this? Seems that not. 
* Transport failures during session. Definitely the most important point to proceed with reconnection (if it is configured)
* Wamp failures during session. Seems that this situations should not lead to reconnection. As they happens when something is broken on business logic side (like using feature that is not announced, using serializer that is not supported etc), so autoreconnection won't automatically fix them.

**Comment by konsultaner** [comment]
With auto healing I basically mean either:

- Transport failures during session. Definitely the most important point to proceed with reconnection (if it is configured)

or 

- A router somehow crashes and the devops environment restarts it automatically. <- simple explaination

I totally agree with:

- Wamp failures during session. Seems that this situations should not lead to reconnection. As they happens when something is broken on business logic side (like using feature that is not announced, using serializer that is not supported etc), so autoreconnection won't automatically fix them.
- Wamp abort (due to auth or other reasons) before session is established. Wamp error means that server processed client messages and get to some point that resolves in abort. Does auto reconnection fix this? Seems that not.

> Yeah! Right now lib operates only on transport state to take decision: reconnect or not.

seems like I have to add some more logic to that. I think will will add this during the week. I need the connectanum-dart for one of my projects too. So finally will have a real live application myself. I will have to add some stuff like auto-resubscribe and other things.

Maybe you take this part off this pull request and add an issue I will work on? Or if you are willing to contribute, feel free to add this logic!

**Comment by KSDaemon** [comment]
Okay! Let me then close this PR, I'll make a new one without these changes, also I'll file a new issue about `better reconnection handling` for you. 
In the meantime i'll continue my work on `Payload Passthru mode` (aka PPT) so look for a new PR later :)
And next big thing is End-2-End Encryption (aka e2ee). So stay tuned! 

May be you can add me as a contributor? Just for a convenient way of creating issues, reviews etc?

**Comment by KSDaemon** [comment]
Closing in favor of https://github.com/konsultaner/connectanum-dart/pull/49

**Pull request closed without merging**